src/core/ext/filters/client_channel/resolver/polling_resolver.cc

```diff
@@ -126,44 +130,67 @@ void PollingResolver::OnRequestCompleteLocked(Result result) {
   if (!shutdown_) {
-    if (result.service_config.ok() && result.addresses.ok()) {
-      // Reset backoff state so that we start from the beginning when the
-      // next request gets triggered.
-      backoff_.Reset();
-    } else {
-      if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
-        gpr_log(GPR_INFO,
-                "[polling resolver %p] resolution failed (will retry): "
-                "address status \"%s\"; service config status \"%s\"",
-                this, result.addresses.status().ToString().c_str(),
-                result.service_config.status().ToString().c_str());
-      }
-      // Set up for retry.
-      // InvalidateNow to avoid getting stuck re-initializing this timer
-      // in a loop while draining the currently-held WorkSerializer.
-      // Also see https://github.com/grpc/grpc/issues/26079.
-      ExecCtx::Get()->InvalidateNow();
-      Timestamp next_try = backoff_.NextAttemptTime();
-      Duration timeout = next_try - ExecCtx::Get()->Now();
-      GPR_ASSERT(!have_next_resolution_timer_);
-      have_next_resolution_timer_ = true;
-      if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
-        if (timeout > Duration::Zero()) {
-          gpr_log(GPR_INFO, "[polling resolver %p] retrying in %" PRId64 " ms",
-                  this, timeout.millis());
-        } else {
-          gpr_log(GPR_INFO, "[polling resolver %p] retrying immediately", this);
-        }
-      }
-      Ref(DEBUG_LOCATION, "next_resolution_timer").release();
-      GRPC_CLOSURE_INIT(&on_next_resolution_, OnNextResolution, this, nullptr);
-      grpc_timer_init(&next_resolution_timer_, next_try, &on_next_resolution_);
+    if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+      gpr_log(GPR_INFO,
+              "[polling resolver %p] returning result: "
+              "addresses=%s, service_config=%s",
+              this,
+              result.addresses.ok()
+                  ? absl::StrCat("<", result.addresses->size(), " addresses>")
+                        .c_str()
+                  : result.addresses.status().ToString().c_str(),
+              result.service_config.ok()
+                  ? (*result.service_config == nullptr
+                         ? "<null>"
+                         : std::string((*result.service_config)->json_string())
+                               .c_str())
+                  : result.service_config.status().ToString().c_str());
     }
+    GPR_ASSERT(result.result_health_callback == nullptr);
+    RefCountedPtr<PollingResolver> self =
+        Ref(DEBUG_LOCATION, "result_health_callback");
+    result.result_health_callback = [self =
+                                         std::move(self)](absl::Status status) {
+      self->GetResultStatus(std::move(status));
+    };
     result_handler_->ReportResult(std::move(result));
   }
   Unref(DEBUG_LOCATION, "OnRequestComplete");
 }
 
+void PollingResolver::GetResultStatus(absl::Status status) {
```

Comment:
Do we have a guarantee that an LB policy's `UpdateLocked` method won't be called between the time that `ReportResult` is called and `GetResultStatus` is called?

(from looking at the code, I think this won't happen, but I'm not sure if we have a more formal guarantee)

Reply:
I'm not sure I understand this question.  The LB policy's `UpdateLocked()` method will *always* be called between when the resolver calls `ReportResult()` and when the result-health callback is invoked.  It has to be that way, because the result of the `UpdateLocked()` call is what determines the status to be passed to the result-health callback.

Reply:
Sorry! I meant to ask what's our guarantee that an LB policy's `UpdateLocked` method won't cause `RequestReResolutionLocked` to be invoked synchronously, i.e. before the health check callback is invoked?

Reply:
Ah, okay -- good question!

It looks like this PR did actually add such a case in pick_first.  I could change pick_first to avoid that, but in the future, when we have more async cases for this, it probably won't be possible to do so.  So instead, I've changed `PollingResolver` to handle that case properly, basically by deferring the re-resolution until after the result-health callback has been invoked.

Thanks for catching this!